data/lua/wml/harm_unit.lua

```diff
@@ -169,10 +195,15 @@ function wml_actions.harm_unit(cfg)
 				and wesnoth.sides.is_enemy( unit_to_harm.side, harmer.side )
 			then
 				if kill ~= false and unit_to_harm.hitpoints <= 0 then
-					harmer.experience = harmer.experience + calc_xp( unit_to_harm.level )
-				else
```

Comment:
It's confusing to have both `kill` and `xp_mode.kill`, this `else` block is for "defending unit survives", not something based on the `xp_mode`.

Reply:
Honestly I dont know what to change it to and for me kill ~= false pretty much means not dead unit but I understand why you can find that confusing to have kill two times.

Reply:
There's *three* definitions of kill/death here, and from the test failures I guess you're missing significance of the third definition for this if...else block. The third definition is that the unit is only dead when `unit_to_harm.hitpoints <= 0`.

`kill ~= false` means the unit might be dead. Checking the value of it here is looking like redundant logic, because the code above this has already reduced the damage if `kill == false`, thus the logic should be able to just check `unit_to_harm.hitpoints <= 0`.